### DISCUSSION TITLE
Can't filter by a relation

### DISCUSSION THREAD

**Discussion by nickolasdeluca** [prisma/prisma#14523] [Q&A] [answered]
Been using Prisma for a little while now, loving it so far.

So far, I could solve all my issues reading through the doc, which is very intuitive and helpful.
But one thing I can't seem to make it work is `filtering` a `relation` in a `findMany` query.

Some information that might be useful:
These are the versions of my Prisma dependencies:
``` json
"@prisma/client": "^4.1.0",
"prisma": "^4.1.0"
```

Here is the code fragment I pulled together to try to achieve this goal:

``` Javascript
const cities = await prisma.addresses_cities.findMany({
  select: {
    ibge_code: true,
    name: true,
    addresses_states: {
      select: {
        id: true,
        name: true,
        initials: true,
      },
      where: { id: 35 },
    },
    created_at: true,
    updated_at: true,
  },
});
```
The goal is to filter all `cities` which belong to `state` 35.
`addresses_states` is the foreign entity.

I tried my best to mimic exactly what is happening in the docs, described [here](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#filter-a-list-of-relations), but the client throws an error:
```
Invalid `prisma.addresses_cities.findMany()\` invocation in

  45 const prisma = new PrismaClient();
  46 
  47 try {
→ 48   const cities = await prisma.addresses_cities.findMany({
         select: {
           ibge_code: true,
           name: true,
           addresses_states: {
             select: {
               id: true,
               name: true,
               initials: true
             },
             where: {
             ~~~~~
               id: 35
             }
           },
           created_at: true,
           updated_at: true
         }
       })

Unknown arg `where` in select.addresses_states.where for type addresses_states. Did you mean `select`? Available args:
type addresses_states {

}
```
As far as I can tell, there is no `where` argument inside the relation, which should exist if you take a look in the docs...

Am I missing something here?

**Comment by austincrim** [comment, marked as answer]
Hey @nickolasdeluca, glad to hear you are enjoying Prisma so far!

Could you share your full Prisma schema? That will help me accurately reproduce this, if possible.

At first glance, I think your query should work as written, but I will need to dig in further.

**Reply by nickolasdeluca** [reply]
here you go:
``` prisma
model addresses {
  cep              Int              @id
  street           String?          @db.VarChar(100)
  district         String?          @db.VarChar(100)
  city_ibge_code   String           @db.VarChar(30)
  created_at       DateTime?        @db.Timestamp(0)
  updated_at       DateTime?        @db.Timestamp(0)
  addresses_cities addresses_cities @relation(fields: [city_ibge_code], references: [ibge_code], map: "fk_address_city_ibge_code")
  companies        companies[]
}

model addresses_cities {
  ibge_code        String           @id @db.VarChar(30)
  name             String           @db.VarChar(100)
  state_id         Int
  created_at       DateTime?        @db.Timestamp(0)
  updated_at       DateTime?        @db.Timestamp(0)
  addresses_states addresses_states @relation(fields: [state_id], references: [id], map: "fk_state_city")
  addresses        addresses[]
  companies        companies[]
}

model addresses_states {
  id               Int                @id
  name             String             @db.VarChar(100)
  initials         String             @db.VarChar(3)
  created_at       DateTime?          @db.Timestamp(0)
  updated_at       DateTime?          @db.Timestamp(0)
  addresses_cities addresses_cities[]
}
```

**Reply by austincrim** [reply]
> The goal is to filter all cities which belong to state 35.
addresses_states is the foreign entity.

If this is your goal, have you tried an approach like this?
```ts
const cities = await prisma.addresses_cities.findMany({
  select: {
    ibge_code: true,
    name: true,
    addresses_states: {
      select: {
        id: true,
        name: true,
        initials: true
      }
    },
    created_at: true,
    updated_at: true
  },
  where: {
    addresses_states: {
      id: 35
    }
  }
})
```

**Reply by nickolasdeluca** [reply]
You sir are a genius, but it is weird, it doesn't look like whats written in the docs, do you agree?

**Reply by austincrim** [reply]
The nested `where` option shown in the docs works in some scenarios, but apparently not all. I'm going to do some more investigation on this and I will provide any relevant updates here.

**Reply by nickolasdeluca** [reply]
You're totally right, what I meant is that the docs appear to be outdated... here is an example:
This is the sample shown in the docs, I've marked with an * where the `where` argument for filtering relations is located
``` javascript
const result = await prisma.user.findMany({
  where: {
    email: {
      contains: 'prisma.io',
    },
  },
  select: {
    posts: {
      * where: {
        published: false,
      },
      orderBy: {
        title: 'asc',
      },
      select: {
        title: true,
      },
    },
  },
})
```
Following your suggestion, this is how the code is supposed to be:
Again, I've marked with * where my `where` argument is located
``` javascript
const result = await prisma.user.findMany({
  * where: {
    email: {
      contains: 'prisma.io',
    },
   * posts: {
     published: false,
   }
  },
  select: {
    posts: {
      orderBy: {
        title: 'asc',
      },
      select: {
        title: true,
      },
    },
  },
})
```
My best guess is that the `where` argument was indeed located inside the nested object in earlier versions, but was changed recently. But that's just a guess... maybe I'm completely wrong. haha
I'll try to find some time and do some testing, I think the `orderBy` argument has also changed places in the code, because when I tested a couple of days ago, I also got thrown an error that the `orderBy` argument was not valid.

**Comment by hdaws** [comment]
@austincrim I have hit this same issue in a fairly complicated schema, where nesting the where clauses with their selects would be much better (ie as the documentation states).  I'm not sure the proposed workaround will suffice for all our queries where this capability is needed. I'm still evaluating how moving the where up will impact lists and other scenarios. Is there any update on this?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
